Refactoring DataVisualization to classes

mathparracho opened the pull request:
## Summary

This PR refactors the `data_visualization` module so that each plot is implemented as its own class, while keeping the existing `DataVisualization` static API working.

Closes #45

## Changes

- Added `BaseVisualizer` in `base_visualizer.py` with a common `plot()` interface.
- Extracted shared helpers into `plot_utils.py` (`save_plot`, `create_subplot_grid`).
- Created one visualizer class per plot, each in its own file:
  - `PlotSeries` (`plot_series.py`)
  - `PlotMultipleSeries` (`plot_multiple_series.py`)
  - `CorrelationHeatmap` (`correlation_heatmap.py`)
  - `PlotFFT` (`plot_fft.py`)
  - `SeasonalDecompositionPlot` (`seasonal_decomposition.py`)
  - `WaveletSpectrogramPlot` (`wavelet_spectrogram.py`)
- Updated `DataVisualization` (in `plots.py`) to act as a façade that:
  - exposes the same static methods as before, and
  - internally delegates to the new visualizer classes.
- Kept `three_w_chart.py` unchanged.

## Backwards compatibility

- All existing public entry points are preserved:
  - `DataVisualization.plot_series(...)`
  - `DataVisualization.plot_multiple_series(...)`
  - `DataVisualization.correlation_heatmap(...)`
  - `DataVisualization.plot_fft(...)`
  - `DataVisualization.seasonal_decompose(...)`
  - `DataVisualization.plot_wavelet_spectrogram(...)`
- `DataVisualization._save_plot` and `DataVisualization.create_subplot_grid`
  remain available and now wrap the helpers in `plot_utils.py`.


By creating this pull request, I confirm that I have read and fully accept and agree with one of the **Petrobras' Contributor License Agreements (CLAs)**: 

* ICLA: [Individual Contributor License Agreement](../blob/main/clas/individual_cla.md) on behalf of **only yourself**;
* CCLA: [Corporate Contributor License Agreement](../blob/main/clas/corporate_cla.md) on behalf of **your employer**.

Our CLAs are based on the *Apache Software Foundation's CLAs*:

* ICLA: [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf)
* CCLA: [Corporate Contributor License Agreement](https://www.apache.org/licenses/cla-corporate.pdf)

rafaelpadilla commented on toolkit/ThreeWToolkit/data_visualization/plot_fft.py:
```diff
@@ -0,0 +1,61 @@
+        if clean_series.empty:
+            raise ValueError("Series contains only NaN values")
+
+        N = len(clean_series)
```
More representative name for the variable `N`. Let's keep capitalized variables to constants.
```suggestion
        num_samples = len(clean_series)
```

rafaelpadilla commented on toolkit/ThreeWToolkit/data_visualization/plot_fft.py:
```diff
@@ -0,0 +1,61 @@
+            T = 1.0
+            freq_unit = "Cycles per Sample"
+        else:
+            T = 1.0 / self.sample_rate
```
And here:
```suggestion
            sample_period = 1.0 / self.sample_rate
```

rafaelpadilla commented on toolkit/ThreeWToolkit/data_visualization/plot_multiple_series.py:
```diff
@@ -0,0 +1,60 @@
+        colors = [cmap(i) for i in range(len(self.series_list))]
+
+        plotted_any = False
+        for i, (series, label) in enumerate(zip(self.series_list, self.labels)):
```
If lengths are guaranteed equal, this is fine.
Let's use `strict=True`  for an extra safety.

```suggestion
        for i, (series, label) in enumerate(zip(self.series_list, self.labels, strict=True)):
```

rafaelpadilla commented on toolkit/ThreeWToolkit/data_visualization/plot_multiple_series.py:
```diff
@@ -0,0 +1,60 @@
+        for i, (series, label) in enumerate(zip(self.series_list, self.labels)):
+            ax.plot(
+                series.index,
+                np.asarray(series.values),
```
I believe `np.asarray(series.values)` is redundant. `series.values` is already a NumPy array.

```suggestion
                series.values,
```

rafaelpadilla commented on toolkit/ThreeWToolkit/data_visualization/correlation_heatmap.py:
```diff
@@ -0,0 +1,61 @@
+            fig, ax = plt.subplots(figsize=figsize)
+        else:
+            fig = cast(Figure, ax.get_figure())
+
```
For the other plots you checked if the df contains only NaN to prevent an empty plot.
Maybe it would be good to do the same here:

```suggestion
        clean_series = self.df_of_series.dropna()
        if clean_series.empty:
            raise ValueError("Series contains only NaN values")
```

rafaelpadilla commented on toolkit/ThreeWToolkit/data_visualization/plot_multiple_series.py:
```diff
@@ -0,0 +1,60 @@
+
+        plotted_any = False
+        for i, (series, label) in enumerate(zip(self.series_list, self.labels)):
+            ax.plot(
```
Before `ax.plot` it would be good to check if series contains only NaN values, as done for another plot:

```
clean_series = series.dropna()
if clean_series.empty:
    continue
```

rafaelpadilla commented on toolkit/ThreeWToolkit/data_visualization/plot_series.py:
```diff
@@ -0,0 +1,62 @@
+            nan_dates = self.series.index[self.series.isna()]
+            for date in nan_dates:
+                ax.axvline(x=date, color="red", linestyle="--", alpha=0.7, linewidth=1)
+
```
Maybe is worth checking if the plot would be empty. If so, similarly as suggested for the other plots, we could raise an error here.

rafaelpadilla commented on toolkit/ThreeWToolkit/data_visualization/correlation_heatmap.py:
```diff
@@ -0,0 +1,61 @@
+        self.df_of_series = df_of_series
+        self.kwargs = kwargs
+
+    def plot(self, ax: Axes | None = None) -> tuple[Figure, Axes]:
```
Docstring missing.

rafaelpadilla commented on toolkit/ThreeWToolkit/data_visualization/plot_multiple_series.py:
```diff
@@ -0,0 +1,60 @@
+        self.ylabel = ylabel
+        self.plot_kwargs = plot_kwargs
+
+    def plot(self, ax: Axes | None = None) -> tuple[Figure, Axes]:
```
Docstring missing.

rafaelpadilla replied:
docstring

rafaelpadilla commented on toolkit/ThreeWToolkit/data_visualization/plot_series.py:
```diff
@@ -0,0 +1,62 @@
+        ylabel: str,
+        overlay_events: bool = False,
+        **plot_kwargs,
+    ) -> None:
```
docstring missing

rafaelpadilla commented on toolkit/ThreeWToolkit/data_visualization/plot_series.py:
```diff
@@ -0,0 +1,62 @@
+        self.overlay_events = overlay_events
+        self.plot_kwargs = plot_kwargs
+
+    def plot(self, ax: Axes | None = None) -> tuple[Figure, Axes]:
```
docstring missing

rafaelpadilla commented on toolkit/ThreeWToolkit/data_visualization/seasonal_decomposition.py:
```diff
@@ -0,0 +1,85 @@
+        series: pd.Series,
+        model: str = "additive",
+        period: int | None = None,
+    ) -> None:
```
docstring

rafaelpadilla commented on toolkit/ThreeWToolkit/data_visualization/seasonal_decomposition.py:
```diff
@@ -0,0 +1,85 @@
+        self.model = model
+        self.period = period
+
+    def plot(self, ax: Axes | None = None) -> tuple[Figure, Axes]:
```
docstring

mathparracho commented:
Hi @rafaelpadilla, thank you for the detailed review.

I went through all the points you mentioned and applied the corresponding updates across the visualization module. In summary:

- Added missing docstrings to the classes, constructors, and plot methods.

- Fixed variable names that were incorrectly capitalized.

- Removed unnecessary np.array() conversions when the inputs were already NumPy arrays.

- Standardized empty-series and NaN-only checks across all plotters, raising consistent ValueError messages.

Because of these changes to the new validations for empty and NaN-only inputs, I had to adjust/extend the logic of some unit tests to reflect the updated expected behavior.

Please let me know if you'd like me to refine anything further.
